Rework the output website elements and drop python 2

jhkennedy opened the pull request:
currently, to produce the website as viewed in the browser, LIVVkit relies on a collection of Javascript functions that parse associated JSON files to dynamically create HTML. This is overly complicated as LIVVkit websites are static and makes customizing LIVVkit display elements hard (you have to edit the Javascript well away from where you create elements). 

This PR reworks the elements to be produced from customizable HTML templates through jinja2, and significantly simplifies both working with the elements and the required Javascript for the website. 

Additionally, as part of this effort, working with the classes and templates is significantly easier without trying to support python 2. Because Python 2 EOL is already here, this also drops python 2 support, moving the supported and tested python versions to 3.6 and up. 

Because of the significance of these changes, develop should now be considered to target the LIVVkit 3.0.0 release (version number bumped accordingly) and this major API change will break associated packages like LEX and EVV without updates. 

Fixes #30 
Fixes #19 
Fixes #20 
Closes #31 (obsolete)

jhkennedy committed:
Define a base element class for new-style elements

This begins the conversion to each each LIVVkit element providing a
_repr_STYLE function, and the LIVVkit output will be built based on that
style.

Current styles include JSON, HTML, LaTeX -- LIVVkit will output an HTML
website and/or a LaTeX document (uncompiled). Along side the HTML and/or
LaTeX, a JSON data representation will be created.

Ideally, with just the JSON, LIVVkit should be able to create an HTML
website, LATEX document, and/or any future provided styles.

jhkennedy committed:
Convert image element from function to Class

LIVVkit produces correct images on output website, but doesn't make use
of the _repr_html function yet

jhkennedy committed:
Convert raw html element from function to class

LIVVkit will produce the correct html on the output website, but doesn't
make use of the _repr_html method yet

jhkennedy committed:
Update package structure for new classed elements

jhkennedy committed:
Remove make_tex and TexHelper utilities

These were being developed to provide LaTeX output by converting the
JSON files from an output website, or converting an element dictionary
to LaTeX. However, because we are adopting elements as classes (see #20)
, a `_repr_latex` method is being provided for all elements,
depreciating these.

closes #14

jhkennedy committed:
Remove forgotten raw_html function

jhkennedy committed:
make html and latex jinja2 environments private

jhkennedy committed:
Convert error element to class

LIVVkit produces correct error messages on output website, but doesn't make use
of the _repr_* functions yet

Commit:
Fix summary table error styling (PR #52)

Reset style for each summary table element

Fixes #10

jhkennedy committed:
Convert file_diff element to class-style element

LIVVkit produces the correct config file differences on the output
website but doesn't yet utilize the class as a class.

jhkennedy committed:
Convert gallery element to class-style element

LIVVkit produces the correct image galleries on the output website, but
does not yet take full advantage of the class methods.

jhkennedy committed:
Better method TypeError message for BaseElement class

jhkennedy committed:
Some formatting cleanup

jhkennedy committed:
livv bubble as raw string to prevent attempted escapes

jhkennedy committed:
Add docstring spec to docs and add document base element

jhkennedy committed:
Convert table element to class-style element

LIVVkit produces the correct tables on the output website, but
does not yet take full advantage of the class methods.

Also, adds Pandas DataFrame as a supported data type for initializing a
table element.

jhkennedy committed:
Convert vhtable element to class-style element

The vhtable (table with vertical and horizontal headers) element has
been subsumed into the Table class with the addition of an index option.
if index=True, the header column will either use the Pandas DataFrame
index or number the table rows. If index=False (default) you'll get a
normal table without a header column. If index=Iterable then the header
column will use the iter-values to label the table rows.

jhkennedy committed:
Properly order pytest asserts in elements tests

jhkennedy committed:
Convert vtable element to class-style element

The vtable (table with vertical headers) element has
been subsumed into the Table class with the addition of an transpose option.
if transpose=True, the columns will be transposed to rows (index across
the top header row if present and normal headers in first column)

jhkennedy committed:
Convert bit_for_bit element to class-style element

The bit-for-bit table (table with bit-for-bit difference plots in final
column) element has been converted to a BitForBit class. LIVVkit
produces the correct b4b tables on the output website, but does not yet
take full advantage of the class methods.

Also adds a B4BImage and a NAImage which take place of the bit-for-bit
difference plots (of nothing-plot) or stand is as a placeholder if the
bit-for-bit comparison can't be run (e.g., variable missing from either
test or reference data), respectively.

Further:
* Does some cleanup of the CSS while ading styling for the images
  in the bit-for-bit table.
* Adds the LIVVkit favicon.ico to all the templates (when'd that
  dissapear?).
* Removes unused common.js functions
* Fixes Gallery Type
* Fixes the URL generation/path specification for Image elements

jhkennedy committed:
Fix incorrect number of columns error in table latex

jhkennedy committed:
Remove obsolete common.js functions

The various draw table functions have been removed and the remaing draw
table function just uses the _repr_html() output stored in the Data
variable.

Also, fix mistaken access of Data variable in table template

jhkennedy committed:
Use actual JSON representation of sub-elements for _repr_json

jhkennedy committed:
Define a CompositeElement for elements that contain elements

Moves the Gallery and BitForBit classes to inherit from CompositeElement
and unifies where they store sub-elements

jhkennedy committed:
Convert section element to class-style element

The section element has been converted to a Section class.  LIVVkit
produces the correct sections on the output website, but does not yet
take full advantage of the class methods.

The verification component of LIVVkit now almost entirely uses the class
style elements, except for summarizing the results.

jhkennedy committed:
Fix missing lightbox for b4b table images

jhkennedy committed:
Make elements a subpackage of livvkit

jhkennedy committed:
PEP8 and remove some PyCharm inspection hints

jhkennedy committed:
Remove errant div-closing HTML tags (I think?)

These tags appear to be unconnected to an opening div tag, but it looks
like web browsers automatically clean this up? I couldn't actually tell
a difference between the webpages before and after. However, non of the
templates leave a hanging div so I'm pretty sure this is correct.

jhkennedy committed:
Simplify error element Javascript

jhkennedy committed:
Update LIVVkit tab elements to new Class style elements

The tab element dictionary has been converted to a Tabs class.  LIVVkit
produces the correct tabs on the output website, but does not yet
take full advantage of the class methods.

In order to do this, a new NamedCompositeElement base class class was
introduced which stores a composite name and a list of contained
elements in a dictionary.

jhkennedy committed:
Remove unused (in livv or lex) Book element

jhkennedy committed:
Update page element to Class and add serial option

The page element dictionary has been converted to a Page class.  LIVVkit
produces the correct pages on the output website, but does not yet take
full advantage of the class methods.

In order to do this, the summary dictionaries had to be handled a little
differently and are returned from run_suite component functions instead
of written to a shared memory dictionary (better practices anyways) and
the multiprocessing method has switched from starting a list of
processes to running them within a pool of workers.

A command line option for `-p/--pool-size` was added so the user can
select the size of the processing pool, where a pool size of zero causes
a serial run. This changes the `-p` option, removing the under utilized
`-p/--publish` option that added publication quality figures because
that methodology wasn't automatic or easy for users to implement.

Similar functionality would be better used through a figure option which
allows you to specify figure properties automatically (though this will
be hard to implement in LEX as users are free to use matplotlib or not).

jhkennedy committed:
Fix navigation and element JSON representation

Navigation was broken in part because the JSON representation of
composite elements was incorrect (a string of elements JSON, instead of
just JSON).

This also removes all unused Javascript and simplifies the DrawNav
Javascript function.

jhkennedy committed:
Minor PEP8 Fixes

jhkennedy committed:
Remove obsolete element attributes

jhkennedy committed:
Fix templates in MANIFEST.in

jhkennedy committed:
Drop py27 and add py38 in tox

jhkennedy committed:
Drop py27 and py35, and add py38 in Travis-CI

jhkennedy committed:
Specify python version requirements in setup.py

jhkennedy committed:
Drop python 2; remove six and future imports

jhkennedy committed:
Remove unecessary netcdf module to reduce confusion with netCDF package

jhkennedy committed:
Bump version to next target -- 3.0.0

mkstratos commented on livvkit/bundles/CISM_glissade/verification.py:
```diff
@@ -52,14 +50,9 @@ def parse_log(file_path):
     """
     if not os.path.isfile(file_path):
-        return elements.error("Output Log", "Could not open file: " + file_path.split(os.sep)[-1])
+        return elements.Error("Output Log", "Could not open file: " + file_path.split(os.sep)[-1]).__dict__
```
This returns a dict rather than the `elements.Error` Element, is that correct?

jhkennedy replied:
Ooh, good catch.

mkstratos reviewed (changes requested):
Aside from the one (possible) change in `livvkit/bundles/CISM_glissade/verification.py`, all looks good to me. Though, does `tests\conftest.py` ever get called when testing?

jhkennedy committed:
Fix return of Error Element

jhkennedy committed:
Update documentation for LIVVKit 3.0 (except elements)

jhkennedy committed:
Python is a name

jhkennedy committed:
Add references to Page element

Update Evans et al. (2019) bibtex for publication.

jhkennedy committed:
Eliminate hard-coded paths in test_el_page_json test

jhkennedy committed:
Add docstrings to all LIVVkit elements

jhkennedy committed:
Fix FileDiff element docstring

jhkennedy committed:
Better performance summary mean diff styling

jhkennedy commented:
@mkstratos I updated docs to reflect all these changes (and the changes in dev too). Once you're happy I'll do the merge. 

as for `conftest.py`, that's used by `pytest` when it's run -- it sets up some shared methods that will be available to all the tests. See:
* https://docs.pytest.org/en/2.7.3/plugins.html?highlight=re
* https://stackoverflow.com/questions/34466027/in-pytest-what-is-the-use-of-conftest-py-files

mkstratos reviewed (approved):
Joe, all looks good to me. Can you tag any other issues this will close?

jhkennedy committed:
Resolve conflicts with develop for PR #53

* upstream/develop:
  Reset style for each summary table element, fixing issue 10
  Remove repeated CI job
  Update setup.py for development extras
  Add requests package for coverage on Travis CI
  Fix coverage report on Travis
  Add deepsource analyses

Conflicts:
	.travis.yml
	livvkit/resources/js/common.js

jhkennedy committed:
Unrestrict pybtex version

latest is passing all tests; unclear why it was originally restricted

jhkennedy committed:
Fix some Deepsource inspection issues and remove PyCharm inspection comments

jhkennedy committed:
Silence critical Deepsource.io jinja2 template issue BAN-B701

jhkennedy commented:
Alright, all issues tagged and closed that this one fixes, plus I added some comments on other issues. 

Couple questions: 

1. Is this what we want to release for LIVVkit 3.0? 
2. If so, do you want:
    1. Me to do all the release steps? (merge to master, pypi, conda-forge)
    2. You to do all the release step?
   
   Either way, I'll write add documentation for how to do it all
